we'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website. The aim of this project is to clean the data and analyze the included used car listings.

In [1]:
import pandas as pd
import numpy as np
autos=pd.read_csv('autos.csv', encoding='Latin-1')

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [4]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
autos.rename(columns={"yearOfRegistration":"registration_year", "monthOfRegistration":"registration_month", "notRepairedDamage":"unrepaired_damage", "dateCreated":"ad_created"}, inplace=True)

In [7]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'registration_year', 'gearbox', 'powerPS', 'model',
       'odometer', 'registration_month', 'fuelType', 'brand',
       'unrepaired_damage', 'ad_created', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [8]:
autos.rename(columns={"yearOfRegistration":"registration_year", "monthOfRegistration":"registration_month", "notRepairedDamage":"unrepaired_damage", "dateCreated":"ad_created", "dateCrawled":"date_crawled", "offerType":"offer_type", "vehicleType":"vehicle_type", "powerPS":"power_ps", "fuelType":"fuel_type", "notRepariedDamage":"not_repaired_damage", "nrOfPictures":"no_of_pictures", "postalCode":"postal_code", "lastSeen":"last_seen" }, inplace=True)

In [9]:
autos.columns


Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'no_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [10]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  no_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Changed the column names using dataframe.rename attritbute to make them easily readable in camel case.

In [11]:
autos.describe()

registration_year      power_ps  registration_month  no_of_pictures  \
count       50000.000000  50000.000000        50000.000000         50000.0   
mean         2005.073280    116.355920            5.723360             0.0   
std           105.712813    209.216627            3.711984             0.0   
min          1000.000000      0.000000            0.000000             0.0   
25%          1999.000000     70.000000            3.000000             0.0   
50%          2003.000000    105.000000            6.000000             0.0   
75%          2008.000000    150.000000            9.000000             0.0   
max          9999.000000  17700.000000           12.000000             0.0   

        postal_code  
count  50000.000000  
mean   50813.627300  
std    25779.747957  
min     1067.000000  
25%    30451.000000  
50%    49577.000000  
75%    71540.000000  
max    99998.000000

There are no pictures in any of the column, so we can drop the number of pictures column. Price and Odometer columns, which are supposed to be numerical but in the dataframe they are strings. Apart from year of registration, I feel that a new column indicating the age of cars (considering 2019 as the current year) will help decide. 

In [12]:
autos["price"]=autos["price"].str.replace("$","").str.replace(",","").astype(int)

In [13]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price   abtest vehicle_type  registration_year  \
0  privat    Angebot   5000  control          bus               2004   
1  privat    Angebot   8500  control    limousine               1997   
2  privat    Angebot   8990     test    limousine               2009   
3  privat    Angebot   4350  control   kleinwagen               2007   
4  privat    Angebot   1350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  no_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [14]:
autos.describe()

price  registration_year      power_ps  registration_month  \
count  5.000000e+04       50000.000000  50000.000000        50000.000000   
mean   9.840044e+03        2005.073280    116.355920            5.723360   
std    4.811044e+05         105.712813    209.216627            3.711984   
min    0.000000e+00        1000.000000      0.000000            0.000000   
25%    1.100000e+03        1999.000000     70.000000            3.000000   
50%    2.950000e+03        2003.000000    105.000000            6.000000   
75%    7.200000e+03        2008.000000    150.000000            9.000000   
max    1.000000e+08        9999.000000  17700.000000           12.000000   

       no_of_pictures   postal_code  
count         50000.0  50000.000000  
mean              0.0  50813.627300  
std               0.0  25779.747957  
min               0.0   1067.000000  
25%               0.0  30451.000000  
50%               0.0  49577.000000  
75%               0.0  71540.000000  
max               0.0  99998.000000

In [15]:
autos["odometer"]=autos["odometer"].str.replace(",","").str.replace("km","").astype(int)

In [16]:
autos.describe()

price  registration_year      power_ps       odometer  \
count  5.000000e+04       50000.000000  50000.000000   50000.000000   
mean   9.840044e+03        2005.073280    116.355920  125732.700000   
std    4.811044e+05         105.712813    209.216627   40042.211706   
min    0.000000e+00        1000.000000      0.000000    5000.000000   
25%    1.100000e+03        1999.000000     70.000000  125000.000000   
50%    2.950000e+03        2003.000000    105.000000  150000.000000   
75%    7.200000e+03        2008.000000    150.000000  150000.000000   
max    1.000000e+08        9999.000000  17700.000000  150000.000000   

       registration_month  no_of_pictures   postal_code  
count        50000.000000         50000.0  50000.000000  
mean             5.723360             0.0  50813.627300  
std              3.711984             0.0  25779.747957  
min              0.000000             0.0   1067.000000  
25%              3.000000             0.0  30451.000000  
50%              6.000000             0.0  49577.000000  
75%              9.000000             0.0  71540.000000  
max             12.000000             0.0  99998.000000

In [17]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price   abtest vehicle_type  registration_year  \
0  privat    Angebot   5000  control          bus               2004   
1  privat    Angebot   8500  control    limousine               1997   
2  privat    Angebot   8990     test    limousine               2009   
3  privat    Angebot   4350  control   kleinwagen               2007   
4  privat    Angebot   1350     test        kombi               2003   

     gearbox  power_ps   model  odometer  registration_month fuel_type  \
0    manuell       158  andere    150000                   3       lpg   
1  automatik       286     7er    150000                   6    benzin   
2    manuell       102    golf     70000                   7    benzin   
3  automatik        71  fortwo     70000                   6    benzin   
4    manuell         0   focus    150000                   7    benzin   

        brand unrepaired_damage           ad_created  no_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [18]:
autos.rename(columns={"odometer":"odometer_km"})

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine

In [25]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'no_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

You might have noticed that "odometer" column's name is unchanged because I did not change that in place (inplace =True). I changed it in the next cell and printed the column names. Just a point to remember§

In [26]:
autos.rename(columns={"odometer":"odometer_km"}, inplace=True)

In [27]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer_km', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'no_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [28]:
autos.describe()

price  registration_year      power_ps    odometer_km  \
count  5.000000e+04       50000.000000  50000.000000   50000.000000   
mean   9.840044e+03        2005.073280    116.355920  125732.700000   
std    4.811044e+05         105.712813    209.216627   40042.211706   
min    0.000000e+00        1000.000000      0.000000    5000.000000   
25%    1.100000e+03        1999.000000     70.000000  125000.000000   
50%    2.950000e+03        2003.000000    105.000000  150000.000000   
75%    7.200000e+03        2008.000000    150.000000  150000.000000   
max    1.000000e+08        9999.000000  17700.000000  150000.000000   

       registration_month  no_of_pictures   postal_code  
count        50000.000000         50000.0  50000.000000  
mean             5.723360             0.0  50813.627300  
std              3.711984             0.0  25779.747957  
min              0.000000             0.0   1067.000000  
25%              3.000000             0.0  30451.000000  
50%              6.000000             0.0  49577.000000  
75%              9.000000             0.0  71540.000000  
max             12.000000             0.0  99998.000000

In [30]:
autos["odometer_km"].unique()

array([150000,  70000,  50000,  80000,  10000,  30000, 125000,  90000,
        20000,  60000,   5000, 100000,  40000])

In [34]:
autos["odometer_km"].unique().shape

(13,)

In [33]:
type(autos["odometer_km"].unique())

numpy.ndarray

In [35]:
autos["odometer_km"].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

In [36]:
autos["odometer_km"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [37]:
autos["price"].value_counts().shape

(2357,)

In [57]:
autos["price"].value_counts().head(100)

0        1421
500       781
1500      734
2500      643
1000      639
1200      639
600       531
800       498
3500      498
2000      460
999       434
750       433
900       420
650       419
850       410
700       395
4500      394
300       384
2200      382
950       379
1100      376
1300      371
3000      365
550       356
1800      355
5500      340
1250      335
350       335
1600      327
1999      322
         ... 
4999      174
2950      173
2750      169
6900      165
7000      165
8000      161
1499      157
5200      156
1         156
7900      156
3700      155
1550      150
990       147
2250      147
11500     145
8900      144
12500     142
5999      140
3400      137
2499      136
1299      135
9900      135
3950      135
2990      134
100       134
3100      133
4300      132
9999      131
2850      131
13500     130
Name: price, Length: 100, dtype: int64

In [39]:
autos["odometer_km"].value_counts().sort_index(ascending=True)

5000        967
10000       264
20000       784
30000       789
40000       819
50000      1027
60000      1164
70000      1230
80000      1436
90000      1757
100000     2169
125000     5170
150000    32424
Name: odometer_km, dtype: int64

In [56]:
autos["price"].value_counts().sort_index(ascending=False).head(100)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
194000      1
190000      1
180000      1
175000      1
169999      1
169000      1
163991      1
163500      1
155000      1
151990      1
           ..
75900       1
75000       1
74999       2
74900       3
73996       1
73900       1
73500       1
72900       1
72600       1
72500       1
71000       1
70850       1
70000       1
69999       1
69997       1
69993       1
69900       1
69500       1
68900       1
68750       1
68500       1
68300       1
68000       1
67911       1
67000       1
66964       1
66500       1
65990       1
65700       1
65699       1
Name: price, Length: 100, dtype: int64

If I sort the values in reverse order, we see that the values above 350,000 are not valid values, they were just random or outlier values. In the next cell, I used the price value between 500 and 351000. When inspecting closer, the price could start from 1 usd, as ebay is about auction. To avoid excluding such data, I used the value beteen 1 and 351000.

For odometer values, it is not necessary. There is no unreasonably high or invalid values. There are many high mileage cars, which is understandable on an auction site. 

In [54]:
autos["price"].value_counts().sort_index(ascending=True).head(300)

0       1421
1        156
2          3
3          1
5          2
8          1
9          1
10         7
11         2
12         3
13         2
14         1
15         2
17         3
18         1
20         4
25         5
29         1
30         7
35         1
40         6
45         4
47         1
49         4
50        49
55         2
59         1
60         9
65         5
66         1
        ... 
1100     376
1111      39
1112       1
1119       1
1120       2
1149      10
1150     226
1169       1
1170       1
1180       4
1189       1
1190      37
1195       1
1199     126
1200     639
1201       2
1209       1
1212       2
1221       1
1222       3
1234       3
1240       2
1247       1
1249       4
1250     335
1265       1
1270       4
1275       1
1280      12
1285       3
Name: price, Length: 300, dtype: int64

In [55]:
autos[autos["price"].between(500,351000)]

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
31     2016-03-14 16:53:09                          Noch_gut_erhaltenen_C_320   
33     2016-03-15 12:25:26  Audi_A3_2.0_TDI_Sportback_DPF_Kupplung_defekt_...   
34     2016-04-05 07:55:07                            Bmw_530d__navi_tuv03/18   
35     2016-03-29 21:52:56                                        Ford_Mondeo   
36     2016-03-16 23:46:31          Ford_Mondeo_2.0_Turnier_TDCi_DPF_Titanium   
37     2016-03-23 16:47:46                 Seat_Altea_XL_2.0_TDI_CR_DPF_Style   
38     2016-03-21 15:51:10                        Volkswagen_Golf_1.4_Special   
...                    ...                                                ...   
49966  2016-04-02 19:49:19  Citroën_C1_1.0_**Euro4**TÜV_OKT_2017**Scheiten...   
49967  2016-03-12 11:50:17                      VW_Passat_2_0_TDI_comfortline   
49968  2016-04-01 17:49:15                  Mercedes_Benz_190_D_2.5_Automatik   
49969  2016-03-17 18:49:02             Nissan_X_Trail_2.2_dCi_4x4_Sport_m.AHZ   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0

In [64]:
autos=autos[autos["price"].between(1,351000)]

In [65]:
autos.describe()

price  registration_year      power_ps    odometer_km  \
count   48565.000000       48565.000000  48565.000000   48565.000000   
mean     5888.935591        2004.755421    117.197158  125770.101925   
std      9059.854754          88.643887    200.649618   39788.636804   
min         1.000000        1000.000000      0.000000    5000.000000   
25%      1200.000000        1999.000000     71.000000  125000.000000   
50%      3000.000000        2004.000000    107.000000  150000.000000   
75%      7490.000000        2008.000000    150.000000  150000.000000   
max    350000.000000        9999.000000  17700.000000  150000.000000   

       registration_month  no_of_pictures   postal_code  
count        48565.000000         48565.0  48565.000000  
mean             5.782251             0.0  50975.745207  
std              3.685595             0.0  25746.968398  
min              0.000000             0.0   1067.000000  
25%              3.000000             0.0  30657.000000  
50%              6.000000             0.0  49716.000000  
75%              9.000000             0.0  71665.000000  
max             12.000000             0.0  99998.000000

In [66]:
autos["no_of_pictures"].unique()

array([0])

In [67]:
autos["no_of_pictures"].value_counts()

0    48565
Name: no_of_pictures, dtype: int64

In [68]:
autos["seller"].value_counts()

privat        48564
gewerblich        1
Name: seller, dtype: int64

In [69]:
autos["offer_type"].value_counts()

Angebot    48565
Name: offer_type, dtype: int64

As you can see, the number of pictures column is zero everywhere, so there is no additional value in including those columns. In seller information, you can see that they are all private except one commercial seller. Offer type column also has only one value. I decided to remove these columns. 

In [70]:
autos = autos.drop(["no_of_pictures", "seller", "offer_type"], axis=1)

In [71]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   price   abtest vehicle_type  registration_year    gearbox  power_ps  \
0   5000  control          bus               2004    manuell       158   
1   8500  control    limousine               1997  automatik       286   
2   8990     test    limousine               2009    manuell       102   
3   4350  control   kleinwagen               2007  automatik        71   
4   1350     test        kombi               2003    manuell         0   

    model  odometer_km  registration_month fuel_type       brand  \
0  andere       150000                   3       lpg     peugeot   
1     7er       150000                   6    benzin         bmw   
2    golf        70000                   7    benzin  volkswagen   
3  fortwo        70000                   6    benzin       smart   
4   focus       150000                   7    benzin        ford   

  unrepaired_damage           ad_created  postal_code            last_seen  
0              nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

In [72]:
autos[["date_crawled", "ad_created", "last_seen"]][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [74]:
autos["date_crawled"].value_counts().sort_index()

2016-03-05 14:06:30    1
2016-03-05 14:06:40    1
2016-03-05 14:07:04    1
2016-03-05 14:07:08    1
2016-03-05 14:07:21    1
2016-03-05 14:07:26    1
2016-03-05 14:07:40    1
2016-03-05 14:07:45    1
2016-03-05 14:08:00    2
2016-03-05 14:08:05    2
2016-03-05 14:08:27    1
2016-03-05 14:08:42    1
2016-03-05 14:09:02    2
2016-03-05 14:09:05    1
2016-03-05 14:09:20    1
2016-03-05 14:09:22    1
2016-03-05 14:09:38    1
2016-03-05 14:09:46    1
2016-03-05 14:09:56    1
2016-03-05 14:09:57    1
2016-03-05 14:09:58    2
2016-03-05 14:10:18    1
2016-03-05 14:10:20    1
2016-03-05 14:10:46    1
2016-03-05 14:11:03    1
2016-03-05 14:11:05    1
2016-03-05 14:11:14    1
2016-03-05 14:11:15    1
2016-03-05 14:11:25    1
2016-03-05 14:11:40    1
                      ..
2016-04-07 10:36:17    1
2016-04-07 10:36:19    1
2016-04-07 10:36:24    1
2016-04-07 10:36:25    1
2016-04-07 10:36:35    1
2016-04-07 10:36:36    1
2016-04-07 10:36:37    2
2016-04-07 11:06:33    1
2016-04-07 11:36:19    1


In [75]:
autos["date_crawled"].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05 14:06:30    0.000021
2016-03-05 14:06:40    0.000021
2016-03-05 14:07:04    0.000021
2016-03-05 14:07:08    0.000021
2016-03-05 14:07:21    0.000021
2016-03-05 14:07:26    0.000021
2016-03-05 14:07:40    0.000021
2016-03-05 14:07:45    0.000021
2016-03-05 14:08:00    0.000041
2016-03-05 14:08:05    0.000041
2016-03-05 14:08:27    0.000021
2016-03-05 14:08:42    0.000021
2016-03-05 14:09:02    0.000041
2016-03-05 14:09:05    0.000021
2016-03-05 14:09:20    0.000021
2016-03-05 14:09:22    0.000021
2016-03-05 14:09:38    0.000021
2016-03-05 14:09:46    0.000021
2016-03-05 14:09:56    0.000021
2016-03-05 14:09:57    0.000021
2016-03-05 14:09:58    0.000041
2016-03-05 14:10:18    0.000021
2016-03-05 14:10:20    0.000021
2016-03-05 14:10:46    0.000021
2016-03-05 14:11:03    0.000021
2016-03-05 14:11:05    0.000021
2016-03-05 14:11:14    0.000021
2016-03-05 14:11:15    0.000021
2016-03-05 14:11:25    0.000021
2016-03-05 14:11:40    0.000021
                         ...   
2016-04-

In [76]:
autos["date_crawled"].unique().shape

(46882,)

The ads are crawled from dates between 5th March 2016 to 7th April 2016. There are 46882 values and each value is 0.0021 percentage. 

In [77]:
autos["ad_created"].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11 00:00:00    0.000021
2015-08-10 00:00:00    0.000021
2015-09-09 00:00:00    0.000021
2015-11-10 00:00:00    0.000021
2015-12-05 00:00:00    0.000021
2015-12-30 00:00:00    0.000021
2016-01-03 00:00:00    0.000021
2016-01-07 00:00:00    0.000021
2016-01-10 00:00:00    0.000041
2016-01-13 00:00:00    0.000021
2016-01-14 00:00:00    0.000021
2016-01-16 00:00:00    0.000021
2016-01-22 00:00:00    0.000021
2016-01-27 00:00:00    0.000062
2016-01-29 00:00:00    0.000021
2016-02-01 00:00:00    0.000021
2016-02-02 00:00:00    0.000041
2016-02-05 00:00:00    0.000041
2016-02-07 00:00:00    0.000021
2016-02-08 00:00:00    0.000021
2016-02-09 00:00:00    0.000021
2016-02-11 00:00:00    0.000021
2016-02-12 00:00:00    0.000041
2016-02-14 00:00:00    0.000041
2016-02-16 00:00:00    0.000021
2016-02-17 00:00:00    0.000021
2016-02-18 00:00:00    0.000041
2016-02-19 00:00:00    0.000062
2016-02-20 00:00:00    0.000041
2016-02-21 00:00:00    0.000062
                         ...   
2016-03-

46882 ads are created on 76 days between 11th June 2015 to 7th April 2016. 0.002 percentage of ads are created on some days and more than 60% ads were created in the last 3 weeks.

In [96]:
autos["last_seen"].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

The ads were last seen from dates between 5th March 2016 to 7th April 2016. There are 38474 values and 60% of the ads were last seen or removed.

In [79]:
autos["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

In [80]:
autos["registration_year"].value_counts()

2000    3156
2005    2936
1999    2897
2004    2703
2003    2699
2006    2670
2001    2636
2002    2486
1998    2363
2007    2277
2008    2215
2009    2085
1997    1951
2011    1623
2010    1589
2017    1392
1996    1373
2012    1310
1995    1227
2016    1220
2013     803
2014     663
1994     629
2018     470
1993     425
2015     392
1992     370
1990     347
1991     339
1989     174
        ... 
1950       3
9999       3
1800       2
1951       2
1941       2
1934       2
1954       2
1957       2
1955       2
2019       2
1001       1
1939       1
4500       1
1953       1
1111       1
4800       1
5911       1
1943       1
1938       1
1929       1
2800       1
6200       1
4100       1
8888       1
1927       1
9000       1
1948       1
1000       1
1931       1
1952       1
Name: registration_year, Length: 95, dtype: int64

In [99]:
autos["registration_year"].value_counts().sort_index().head(95)

1000       1
1001       1
1111       1
1800       2
1910       5
1927       1
1929       1
1931       1
1934       2
1937       4
1938       1
1939       1
1941       2
1943       1
1948       1
1950       3
1951       2
1952       1
1953       1
1954       2
1955       2
1956       4
1957       2
1958       4
1959       6
1960      23
1961       6
1962       4
1963       8
1964      12
        ... 
2000    3156
2001    2636
2002    2486
2003    2699
2004    2703
2005    2936
2006    2670
2007    2277
2008    2215
2009    2085
2010    1589
2011    1623
2012    1310
2013     803
2014     663
2015     392
2016    1220
2017    1392
2018     470
2019       2
2800       1
4100       1
4500       1
4800       1
5000       4
5911       1
6200       1
8888       1
9000       1
9999       3
Name: registration_year, Length: 95, dtype: int64

There are some invalid values because the first production automobile by Benz was from 1885. We need to take values only from 1910. The ads were crawled latest on 7th April 2016, so the year registered cannot be more than that. We can drop values from 2017 onwards. The number of cars advertised are in 2 digits for years between 1960 to 1987. After that, it is in 3 digits, i.e atleast 100 cars for that model year.  From 1995 onwards, there are more than 1000 cars listed per year till 2012.  

In [104]:
autos=autos[autos["registration_year"].between(1910,2016)]

In [105]:
autos["registration_year"].value_counts(normalize=True).sort_index()

1910    0.000107
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000043
1937    0.000086
1938    0.000021
1939    0.000021
1941    0.000043
1943    0.000021
1948    0.000021
1950    0.000064
1951    0.000043
1952    0.000021
1953    0.000021
1954    0.000043
1955    0.000043
1956    0.000086
1957    0.000043
1958    0.000086
1959    0.000129
1960    0.000493
1961    0.000129
1962    0.000086
1963    0.000171
1964    0.000257
1965    0.000364
1966    0.000471
1967    0.000557
1968    0.000557
          ...   
1987    0.001542
1988    0.002892
1989    0.003727
1990    0.007433
1991    0.007262
1992    0.007926
1993    0.009104
1994    0.013474
1995    0.026285
1996    0.029412
1997    0.041794
1998    0.050620
1999    0.062060
2000    0.067608
2001    0.056468
2002    0.053255
2003    0.057818
2004    0.057904
2005    0.062895
2006    0.057197
2007    0.048778
2008    0.047450
2009    0.044665
2010    0.034040
2011    0.034768
2012    0.028063
2013    0.017202
2014    0.0142

Most sold cars are from year 2000 and then 2005 comes next. More than 70% of the cars sold are from year 2000

In [106]:
autos["brand"].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [128]:
unique_brand_counts={}
brand_list=autos["brand"].unique()
print(brand_list.shape)
type(brand_list)
autos["brand"].value_counts(normalize=True).sort_index()
autos["brand"].value_counts()

    

(40,)


volkswagen        9862
bmw               5137
opel              5022
mercedes_benz     4503
audi              4041
ford              3263
renault           2201
peugeot           1393
fiat              1197
seat               853
skoda              766
nissan             713
mazda              709
smart              661
citroen            654
toyota             593
hyundai            468
sonstige_autos     458
volvo              427
mini               409
mitsubishi         384
honda              366
kia                330
alfa_romeo         310
porsche            286
suzuki             277
chevrolet          266
chrysler           164
dacia              123
daihatsu           117
jeep               106
subaru             100
land_rover          98
saab                77
jaguar              73
daewoo              70
trabant             65
rover               62
lancia              50
lada                27
Name: brand, dtype: int64

In [122]:
autos["brand"].value_counts().index

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'nissan', 'mazda', 'smart',
       'citroen', 'toyota', 'hyundai', 'sonstige_autos', 'volvo', 'mini',
       'mitsubishi', 'honda', 'kia', 'alfa_romeo', 'porsche', 'suzuki',
       'chevrolet', 'chrysler', 'dacia', 'daihatsu', 'jeep', 'subaru',
       'land_rover', 'saab', 'jaguar', 'daewoo', 'trabant', 'rover', 'lancia',
       'lada'],
      dtype='object')

In [129]:
brands_sold=autos["brand"].value_counts()
brands_index=autos["brand"].value_counts().index

In [130]:
brands_index[0]

'volkswagen'

In [131]:
brands_sold[0]

9862

In [132]:
brands_index_sel=brands_index[:20]
brands_sold_sel=brands_sold[:20]

In [134]:
brands_sold_sel

volkswagen        9862
bmw               5137
opel              5022
mercedes_benz     4503
audi              4041
ford              3263
renault           2201
peugeot           1393
fiat              1197
seat               853
skoda              766
nissan             713
mazda              709
smart              661
citroen            654
toyota             593
hyundai            468
sonstige_autos     458
volvo              427
mini               409
Name: brand, dtype: int64

In [136]:
brands_sold.sum()

46681

In [137]:
46681*0.05

2334.05

I have selected the top twenty brands. If I use the criteria of brands that are over 5%, I can have only 6 values. 

In [138]:
brands_index_sel


Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'nissan', 'mazda', 'smart',
       'citroen', 'toyota', 'hyundai', 'sonstige_autos', 'volvo', 'mini'],
      dtype='object')

In [142]:
unique_brand_counts={}
for b in brands_sold_sel:
    print(b)

    


9862
5137
5022
4503
4041
3263
2201
1393
1197
853
766
713
709
661
654
593
468
458
427
409


In [143]:
for b in brands_index_sel:
    print(b)


volkswagen
bmw
opel
mercedes_benz
audi
ford
renault
peugeot
fiat
seat
skoda
nissan
mazda
smart
citroen
toyota
hyundai
sonstige_autos
volvo
mini


In [149]:
unique_brand_counts={}
for b in brands_index_sel:
    unique_brand_counts[b]=int(autos.loc[autos["brand"]==b, "price"].mean())

In [150]:
unique_brand_counts

{'audi': 9336,
 'bmw': 8332,
 'citroen': 3779,
 'fiat': 2813,
 'ford': 3749,
 'hyundai': 5365,
 'mazda': 4112,
 'mercedes_benz': 8628,
 'mini': 10613,
 'nissan': 4743,
 'opel': 2975,
 'peugeot': 3094,
 'renault': 2474,
 'seat': 4397,
 'skoda': 6368,
 'smart': 3580,
 'sonstige_autos': 12338,
 'toyota': 5167,
 'volkswagen': 5402,
 'volvo': 4946}

In [151]:
sorted_brand = sorted((value, key) for (key,value) in unique_brand_counts.items())

In [152]:
sorted_brand

[(2474, 'renault'),
 (2813, 'fiat'),
 (2975, 'opel'),
 (3094, 'peugeot'),
 (3580, 'smart'),
 (3749, 'ford'),
 (3779, 'citroen'),
 (4112, 'mazda'),
 (4397, 'seat'),
 (4743, 'nissan'),
 (4946, 'volvo'),
 (5167, 'toyota'),
 (5365, 'hyundai'),
 (5402, 'volkswagen'),
 (6368, 'skoda'),
 (8332, 'bmw'),
 (8628, 'mercedes_benz'),
 (9336, 'audi'),
 (10613, 'mini'),
 (12338, 'sonstige_autos')]

Sonstige_autos seems expensive but they are sold in less numbers. Those who buy it might be enthusiasts willing to pay more. The same goes for Mini. Of the top 6 selling brands, bmw, mercedes_benz and audi are expensive. Volkswagen is a cheap option, ford is cheaper and opel is the cheapest option.

In [156]:
brand_series=pd.Series(unique_brand_counts)

In [157]:
print(brand_series)

audi               9336
bmw                8332
citroen            3779
fiat               2813
ford               3749
hyundai            5365
mazda              4112
mercedes_benz      8628
mini              10613
nissan             4743
opel               2975
peugeot            3094
renault            2474
seat               4397
skoda              6368
smart              3580
sonstige_autos    12338
toyota             5167
volkswagen         5402
volvo              4946
dtype: int64


In [158]:
brand_df=pd.DataFrame(brand_series, columns=['mean_price'])

In [159]:
brand_df

mean_price
audi                  9336
bmw                   8332
citroen               3779
fiat                  2813
ford                  3749
hyundai               5365
mazda                 4112
mercedes_benz         8628
mini                 10613
nissan                4743
opel                  2975
peugeot               3094
renault               2474
seat                  4397
skoda                 6368
smart                 3580
sonstige_autos       12338
toyota                5167
volkswagen            5402
volvo                 4946

In [160]:
unique_brand_mileage={}
for b in brands_index_sel:
    unique_brand_mileage[b]=int(autos.loc[autos["brand"]==b, "odometer_km"].mean())

In [161]:
unique_brand_mileage

{'audi': 129157,
 'bmw': 132572,
 'citroen': 119694,
 'fiat': 117121,
 'ford': 124266,
 'hyundai': 106442,
 'mazda': 124464,
 'mercedes_benz': 130788,
 'mini': 88105,
 'nissan': 118330,
 'opel': 129310,
 'peugeot': 127153,
 'renault': 128071,
 'seat': 121131,
 'skoda': 110848,
 'smart': 99326,
 'sonstige_autos': 89956,
 'toyota': 115944,
 'volkswagen': 128707,
 'volvo': 138067}

In [162]:
brand_series_mileage=pd.Series(unique_brand_mileage)

In [163]:
brand_series_mileage

audi              129157
bmw               132572
citroen           119694
fiat              117121
ford              124266
hyundai           106442
mazda             124464
mercedes_benz     130788
mini               88105
nissan            118330
opel              129310
peugeot           127153
renault           128071
seat              121131
skoda             110848
smart              99326
sonstige_autos     89956
toyota            115944
volkswagen        128707
volvo             138067
dtype: int64

In [164]:
brand_df["mileage"]=brand_series_mileage

In [165]:
brand_df

mean_price  mileage
audi                  9336   129157
bmw                   8332   132572
citroen               3779   119694
fiat                  2813   117121
ford                  3749   124266
hyundai               5365   106442
mazda                 4112   124464
mercedes_benz         8628   130788
mini                 10613    88105
nissan                4743   118330
opel                  2975   129310
peugeot               3094   127153
renault               2474   128071
seat                  4397   121131
skoda                 6368   110848
smart                 3580    99326
sonstige_autos       12338    89956
toyota                5167   115944
volkswagen            5402   128707
volvo                 4946   138067

Of the top 6 selling brands, bmw, mercedes_benz and audi are expensive. Volkswagen is a cheap option, ford is cheaper and opel is the cheapest option. This analysis did not change considering the fact that mileage variation is less among them (less than 10%).